### PR TITLE
add ANARI_KHR_ARRAY1D_REGION extension

### LINKED ISSUE #16: Add extension for resizing arrays
Some applications can benefit from using a subset of an already allocated ANARI array to minimize internal state change overhead. This proposal seeks to add an extension that uses parameters to control the active sub-region of array elements to be used.

Currently array size is only controllable at array object construction -- in order to change the size of the array data used by an object, an entirely new array object must be constructed. This all but guarantees that new array allocation will occur in most implementations, especially implementations that offload to memory spaces other than the host system's memory (i.e. GPUs). Being able to reuse an existing allocation can be advantageous.

The proposal is to add the `ANARI_KHR_ARRAY1D_REGION` core extension, which adds the following parameter to `ANARIArray1D` respectively:

| Name  | Type     |    Default | Description                                    |
|:------|:---------|-----------:|:----------------------------------------------------|
| region | `UINT64_BOX1`  |          0, *capacity | first (inclusive) and last (exclusive) elements to be used by parent objects |

Array _capacity_ is established on construction -- this is defined as the maximum number of elements the array can contain for the lifetime of the array object. The `begin` and `end` parameters are then clamped to the range [0,capacity] respectively, and warnings should be emitted by the debug layer if `end` is larger than capacity. It is also undefined behavior for `region.lower` >= `region`, meaning implementations should be able to rely on 1) the array containing at least one element in each dimension, and 2) `region.lower`  < `region.upper`.

`UINT64` box types also need to be added to the spec if this proposal is accepted into the standard.

#### EDITS: 

- Change from `begin`/`end` parameters to instead use single `region` parameter
- Only specify the extension for 1D arrays (2/3D is super unlikely, can easily create perf problems)
- Rename extension name from `ANARI_KHR_ARRAY_DYNAMIC_REGION` to `ANARI_KHR_ARRAY1D_REGION`

### PR DESCRIPTION
**Pull request by jeffamstutz**
Closes #16.

### CODE REVIEW

**Comment by jeffamstutz** [comment]
Clarifications have been added based on WG feedback in today's call.

File: chapters/api_concepts.txt [hunk 1]
```diff
@@ -456,6 +456,42 @@ example are the vertex attributes of the
...
+
+The element from `region.lower` is _inclusive_, while `region.upper` is
+_exclusive_. This is intended to mimic C++ begin/end iterators on containers.
+
+If `region` is not set, all elements are used as specified when the array was
+constructed. When `region` is set, it is clamped to the range of the array's
+constructed size respectively and warnings should be emitted by the debug layer
+if `region.upper` is larger than capacity.
+
+It is undefined behavior for `region.lower` >= `region.upper`, meaning
```
Comment (johguenther) [inline chapters/api_concepts.txt:485]: Can we require this stronger? For arrays without this region extension the spec reads "number of elements `numItemsN` must be positive". This would mean `lower` >= `clamped_upper`  is an error (and not undefined behavior).

Alternatively we can also define that
 1. `upper` is clamped from below by `lower+1`
 2. `upper` is clamped from above by "capacity" (like written now)
 3. `lower` is clamped from above by `upper-1`

which would ensure that always `lower<upper` and `size>=1`.

File: chapters/api_concepts.txt [hunk 2]
```diff
@@ -456,6 +456,42 @@ example are the vertex attributes of the
...
+implementations should be able to rely on 1) the array containing at least one
+element, and 2) `region.lower` < `region.upper`.
+
+Mapping an array always returns the full capacity, regardless of the elements
+specified by `region`. Object handles within `region` must always be valid, but
+handles outside of `region` may be invalid. Object arrays can be constructed
```
Comment (johguenther) [inline chapters/api_concepts.txt:491]: This contradicts (or relaxes?) the spec further above
> ANARIArray objects containing object handles increase the ref count of all objects in the array.
Reference counts are updated on creation of the array object and when unmapping the array.

File: chapters/type_table.txt [hunk 3]
```diff
@@ -448,6 +448,22 @@
 |`int32_t[8]`
 |four dimensional 32 bit integer box (lower and upper bound vector)
 
+|`UINT64_BOX1`
+|`uint64_t[2]`
+|one dimensional 64 bit unsigned integer box (lower and upper bounds)
```
Comment (johguenther) [inline chapters/type_table.txt:453]: how about clarifying the bounds here (as well or only here), i.e. `...box (inclusive lower and exclusive upper bounds)`?

File: chapters/api_concepts.txt (hunk 1 above)
Comment (jeffamstutz) [reply chapters/api_concepts.txt:485]: > For arrays without this region extension the spec reads "number of elements numItemsN must be positive"

We should probably reword this to be "number of elements must be >= 1", because the size parameter is unsigned so it will always be positive (by some definition of "positive").

I like the clamping idea. Let's go with that!

File: chapters/api_concepts.txt (hunk 2 above)
Comment (jeffamstutz) [reply chapters/api_concepts.txt:491]: This is orthogonal to object lifetime concerns: valid handles should always have their lifetime managed correctly. This statement in the extension is just guaranteeing that there aren't invalid handles within the range actually in use (i.e. conditional iteration wherever the data is used by the implementation).

Basically this clause lets you create a managed array of objects and only map/write/unmap values into `region` instead of the entire array.

File: chapters/api_concepts.txt [hunk 4]
```diff
@@ -456,6 +456,42 @@ example are the vertex attributes of the
...
+subset of elements. This gives applications the ability to more efficiently
+change the number of elements used by parent objects at the cost of memory size
+efficiency.
+
+Implementations provide this functionality by implementing
+<<core_extensions_array1d_region, extension `KHR_ARRAY1D_REGION`>>, which adds
+the following parameter to `ANARIArray1D`:
+
+.Additional parameters accepted by `ANARIArray1D` with <<core_extensions_array1d_region, extension `KHR_ARRAY1D_REGION`>>.
+[cols="<,<,>,<4",options="header,unbreakable"]
```
[the remark this answers is not in the document]
Comment (jeffamstutz) [reply chapters/api_concepts.txt]: Fixed in 698971e.

File: chapters/api_concepts.txt (hunk 1 above)
Comment (jeffamstutz) [reply chapters/api_concepts.txt]: Cool, works for me.

File: chapters/api_concepts.txt (hunk 2 above)
Comment (johguenther) [reply chapters/api_concepts.txt:495]: OK, I thought that "ANARIArray objects containing object handles increase the ref count of all objects in the array. Reference counts are updated on creation of the array object and when unmapping the array" _implies_ that _all_ handles need to be valid (on creation and on unmap). Which cannot be true for managed arrays on creation... So, it actually means that `NULL` (invalid) handles are ignored during lifetime management. But using invalid `NULL` handles during rendering would be an error. The region extension allows to reduce the active array size, the used elements. So, strictly speaking, this clarification (that handles outside the region can be invalid) is redundant?

Comment (jeffamstutz) [reply chapters/api_concepts.txt]: Yes, I removed the redundant part in 34a4e80.

File: chapters/type_table.txt (hunk 3 above)
Comment (jeffamstutz) [reply chapters/type_table.txt:453]: Then we should use a different type name, IMO: something like `UINT64_ELEMENT_RANGE`. We allow empty bounds (empty `FLOAT32_BOX3`, that is), so it's weird to me to state that the type disallows this in any context it possibly could be used in.

**Comment by johguenther** [comment]
squash? "squash and merge" is not enabled for this repo.

**Comment by jeffamstutz** [comment]
I just fixed that, should be able to squash now.
